examples/turtlesim/src/turtle_frame.cpp

```diff
@@ -0,0 +1,304 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "turtlesim/turtle_frame.h"
+
+#include <QPointF>
+
+#include <cstdlib>
+#include <ctime>
+
+#define DEFAULT_BG_R 0x45
+#define DEFAULT_BG_G 0x56
+#define DEFAULT_BG_B 0xff
+
+namespace turtlesim
+{
+
+TurtleFrame::TurtleFrame(rclcpp::Node::SharedPtr& node_handle, QWidget* parent, Qt::WindowFlags f)
+: QFrame(parent, f)
+, path_image_(500, 500, QImage::Format_ARGB32)
+, path_painter_(&path_image_)
+, frame_count_(0)
+, id_counter_(0)
+{
+  setFixedSize(500, 500);
+  setWindowTitle("TurtleSim");
+
+  srand(time(NULL));
+
+  update_timer_ = new QTimer(this);
+  update_timer_->setInterval(16);
+  update_timer_->start();
+
+  connect(update_timer_, SIGNAL(timeout()), this, SLOT(onUpdate()));
+
+  nh_ = node_handle;
+  rcl_interfaces::msg::IntegerRange range;
+  range.from_value = 0;
+  range.step = 1;
+  range.to_value = 255;
+  rcl_interfaces::msg::ParameterDescriptor background_r_descriptor;
+  background_r_descriptor.description = "Red channel of the background color";
+  background_r_descriptor.integer_range.push_back(range);
+  rcl_interfaces::msg::ParameterDescriptor background_g_descriptor;
+  background_g_descriptor.description = "Green channel of the background color";
+  background_g_descriptor.integer_range.push_back(range);
+  rcl_interfaces::msg::ParameterDescriptor background_b_descriptor;
+  background_b_descriptor.description = "Blue channel of the background color";
+  background_b_descriptor.integer_range.push_back(range);
+  nh_->declare_parameter("background_r", rclcpp::ParameterValue(DEFAULT_BG_R), background_r_descriptor);
+  nh_->declare_parameter("background_g", rclcpp::ParameterValue(DEFAULT_BG_G), background_g_descriptor);
+  nh_->declare_parameter("background_b", rclcpp::ParameterValue(DEFAULT_BG_B), background_b_descriptor);
+
+  QVector<QString> turtles;
+  turtles.append("ardent.png");
+  turtles.append("bouncy.png");
+  turtles.append("crystal.png");
+  turtles.append("dashing.png");
+  turtles.append("eloquent.png");
+  turtles.append("foxy.png");
+  turtles.append("galactic.png");
+  turtles.append("humble.png");
+  turtles.append("rolling.png");
+
+  /*
+   * QString images_path = (ament_index_cpp::get_package_share_directory("turtlesim") + "/images/").c_str();
+   * AndreasKnoblach 2023-01-15: get_package_share_directory cause crash because
+   *   Environment variable 'AMENT_PREFIX_PATH' is not set or empty
+   * Since this is only needed to determine the images_path, it is replaced by hard coding the path.
+   */
+  // QString images_path{"<complete path to rules_ros2>/examples/turtlesim/images/"}; //inentionally comment to trigger compile error
```

Comment:
ATM, `ament_index_cpp::get_package_share_director` is not yet working properly. Hence the path to the turtle images needs to be configured manually, here.